key_management/src/keystore.rs

```diff
@@ -113,127 +117,387 @@ pub trait KeyStore {
-        }
-    }
+pub enum KeyStoreConfig {
+    Memory(),
+    Persistent(PathBuf),
+    Encrypted(PathBuf, String),
 }
 
-impl KeyStore for MemKeyStore {
-    fn list(&self) -> Vec<String> {
-        self.key_info.iter().map(|(key, _)| key.clone()).collect()
-    }
-
-    fn get(&self, k: &str) -> Result<KeyInfo, Error> {
-        self.key_info.get(k).cloned().ok_or(Error::KeyInfo)
-    }
-
-    fn put(&mut self, key: String, key_info: KeyInfo) -> Result<(), Error> {
-        if self.key_info.contains_key(&key) {
-            return Err(Error::KeyExists);
-        }
-        self.key_info.insert(key, key_info);
-        Ok(())
-    }
+/// Persistent KeyStore in JSON cleartext in KEYSTORE_LOCATION
+#[derive(Clone, PartialEq, Debug, Eq)]
+struct PersistentKeyStore {
+    file_path: PathBuf,
+}
 
-    fn remove(&mut self, key: String) -> Result<KeyInfo, Error> {
-        self.key_info.remove(&key).ok_or(Error::KeyInfo)
-    }
+/// Encrypted KeyStore
+/// Argon2id hash key derivation
+/// XSalsa20Poly1305 authenticated encryption
+/// CBOR encoding
+#[derive(Clone, PartialEq, Debug, Eq)]
+struct EncryptedKeyStore {
+    salt: pwhash::Salt,
+    encryption_key: Arc<secretbox::Key>,
 }
 
-/// KeyStore that persists data in KEYSTORE_LOCATION
-#[derive(Default, Clone, PartialEq, Debug, Eq)]
-pub struct PersistentKeyStore {
-    pub key_info: HashMap<String, KeyInfo>,
-    location: String,
+#[derive(Debug, Error)]
+pub enum EncryptedKeyStoreError {
+    /// Possibly indicates incorrect passphrase
+    #[error("Error decrypting data")]
+    DecryptionError,
+    /// An error occured while encrypting keys
+    #[error("Error encrypting data")]
+    EncryptionError,
+    /// Unlock called without `encrypted_keystore` being enabled in config.toml
+    #[error("Error with forest configuration")]
+    ConfigurationError,
 }
 
-impl PersistentKeyStore {
-    pub fn new(location: String) -> Result<Self, Error> {
-        let loc = format!("{}{}", location, KEYSTORE_NAME);
-        let file_op = File::open(&loc);
-        match file_op {
-            Ok(file) => {
-                let reader = BufReader::new(file);
-                let data: HashMap<String, KeyInfo> = serde_json::from_reader(reader)
-                    .map_err(|e| {
-                        error!("failed to deserialize keyfile, initializing new");
-                        e
-                    })
-                    .unwrap_or_default();
-                Ok(Self {
-                    key_info: data,
-                    location: loc,
-                })
+impl KeyStore {
+    pub fn new(config: KeyStoreConfig) -> Result<Self, Error> {
+        match config {
+            KeyStoreConfig::Memory() => Ok(Self {
+                key_info: HashMap::new(),
+                persistence: None,
+                encryption: None,
+            }),
+            KeyStoreConfig::Persistent(location) => {
+                let file_path = location.join(Path::new(KEYSTORE_NAME));
```

Comment:
We can just pass `KEYSTORE_NAME` to join, as it accepts anything that implements AsRef<Path> and &str does: https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.join